src/text/line.rs

```diff
@@ -586,31 +553,87 @@ impl WidgetRef for Line<'_> {
+                Some(Alignment::Right) => line_width.saturating_sub(area.width),
                 Some(Alignment::Left) | None => 0,
             };
-            x = x.saturating_add(offset);
-            self.to_owned()
+            render_spans(&self.spans, area, buf, 0, offset);
         };
-        for span in &line.spans {
-            let span_width = span.width() as u16;
-            let span_area = Rect {
-                x,
-                width: span_width.min(area.right().saturating_sub(x)),
-                ..area
-            };
-            span.render(span_area, buf);
-            x = x.saturating_add(span_width);
-            if x >= area.right() {
-                break;
+    }
+}
+
+/// Renders all the spans of the line that are after the `span_offset` in the line and that are
+/// visible in the given area (starting rendering at `area.x + buf_x_offset`). `buf_x_offset`
+/// probably should just be used to modify area.x instead of being a parameter, but it's here
+/// because that would have made the above code just a little uglier
+///
+/// Algorithm:
+/// 1. For each span in the line:
+///    1. If the span is completely before the `span_offset`, skip it
+///    2. If the span is partially before the `span_offset`, render the right side of the span
+///    3. If the span is completely after the `span_offset`, render the whole span
+/// 2. Update `buf_x_offset` to account for the width of the rendered span
+/// 3. Update `span_offset` to 0 after the first span is rendered in full
+/// 4. Repeat
+fn render_spans(
+    spans: &[Span],
+    area: Rect,
+    buf: &mut Buffer,
+    mut buf_x_offset: u16,
+    mut span_offset: u16,
+) {
+    for span in spans {
+        let span_width = span.width() as u16;
+        // ignore spans that are completely before the offset
+        if span_offset >= span_width {
+            span_offset -= span_width;
+            continue;
+        }
+        let area = Rect {
+            x: area.x.saturating_add(buf_x_offset),
+            width: area.width.saturating_sub(buf_x_offset),
+            ..area
+        };
+        if area.is_empty() {
+            break;
+        }
+        let span_is_only_partially_visible = span_offset > 0;
+        if span_is_only_partially_visible {
+            // render the partially visible span starting at the offset -> right side
+            use unicode_truncate::UnicodeTruncateStr;
+
+            let limit = span_width - span_offset;
+            let (content, width) = span.content.unicode_truncate_start(usize::from(limit));
+            let width = width as u16;
+
+            let x = limit - width;
+
+            // Empty the cells before as there is a multi width grapheme not rendered before. But as
+            // it is not rendered in the first place (its truncated away) dont reset, set " ".
+            for x in area.x..x {
+                buf.get_mut(x, area.y).set_style(span.style).set_symbol(" ");
```

Comment:
Thinking a bit more about this, I don't think rendering spaces is necessarily the right thing for truncated wide emojis for two reasons:

1. The underlying character of the cell might be a better choice. Consider an example where there is a block, with a border, and we render a title that is "🦀crab" right aligned (assume "/-\" represent the borders)
- 8 => `"/--🦀crab\"`
- 7 => `"/-🦀crab\"`
- 6 => `"/🦀crab\"`
- 5 => `"/ crab\"` wouldn't `/-crab\` be nicer here?
- 4 => `"/crab\"`
- 3 => `"/rab\"`

2. The truncated emoji might be at the end of the first span (left aligned example
- 8 => `"/crab🦀--\"`
- 7 => `"/crab🦀-\"`
- 6 => `"/crab🦀\"`
- 5 => `"/ crab\"` wouldn't `/crab-\` be nicer here?
- 4 => `"/crab\"`
- 3 => `"/cra\"`

Reply:
We can omit this code and add it later again when there seems to be a real need for this. The important part is aligning it to the right when the truncated width is shorter than the space (due to the emoji that doesn't fit)

Reply:
I fixed this up in the latest push.